impeller/scene/importer/vertices_builder.cc

```diff
@@ -0,0 +1,107 @@
+
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/scene/importer/vertices_builder.h"
+
+#include <cstring>
+#include <limits>
+
+#include "impeller/scene/importer/conversions.h"
+#include "impeller/scene/importer/scene_flatbuffers.h"
+
+namespace impeller {
+namespace scene {
+namespace importer {
+
+VerticesBuilder::VerticesBuilder() = default;
+
+std::map<VerticesBuilder::Attribute, VerticesBuilder::AttributeProperties>
+    VerticesBuilder::kAttributes = {
+        {VerticesBuilder::Attribute::kPosition,
+         {.offset_bytes = offsetof(Vertex, position),
+          .size_bytes = sizeof(Vertex::position),
+          .component_count = 3}},
+        {VerticesBuilder::Attribute::kNormal,
+         {.offset_bytes = offsetof(Vertex, normal),
+          .size_bytes = sizeof(Vertex::normal),
+          .component_count = 3}},
+        {VerticesBuilder::Attribute::kTangent,
+         {.offset_bytes = offsetof(Vertex, tangent),
+          .size_bytes = sizeof(Vertex::tangent),
+          .component_count = 4}},
+        {VerticesBuilder::Attribute::kTextureCoords,
+         {.offset_bytes = offsetof(Vertex, texture_coords),
+          .size_bytes = sizeof(Vertex::texture_coords),
+          .component_count = 2}},
+        {VerticesBuilder::Attribute::kColor,
+         {.offset_bytes = offsetof(Vertex, color),
+          .size_bytes = sizeof(Vertex::color),
+          .component_count = 4}}};
+
+void VerticesBuilder::WriteFBVertices(std::vector<fb::Vertex>& vertices) const {
+  vertices.resize(0);
+  for (auto& v : vertices_) {
+    vertices.push_back(fb::Vertex(
+        ToFBVec3(v.position), ToFBVec3(v.normal), ToFBVec4(v.tangent),
+        ToFBVec2(v.texture_coords), ToFBColor(v.color)));
+  }
+}
+
+template <typename SourceType, size_t Divisor>
+void WriteScalars(void* destination,
```

Comment:
Should this be private member of VerticesBuilder, static, or in an anonymous namespace?

Also, looking at how this is called below, I'm not sure what this does. For example, If the Divisor is the max uint8_t, won't the result be all 0?

Reply:
This can just be static. I changed the name to be a bit less ambiguous and added a comment to explain this conversion behavior. I went ahead and also removed the Divisor template param to resolve it inline with a constexpr to make things less confusing/disconnected.

```
/// @brief  Reads a contiguous sequence of numeric components from `source` and
///         writes them to `destination` buffer as 32bit floats. Signed
///         SourceTypes convert to a range of -1 to 1, and unsigned SourceTypes
///         convert to a range from 0 to 1.
```